src/cli/search.rs

```diff
@@ -45,44 +45,50 @@ pub struct Args {
     limit: Option<usize>,
 }
 
-/// fetch packages from `repo_data` based on `filter_func`
-fn search_package_by_filter<F>(
+/// fetch packages from `repo_data` using `repodata_query_func` based on `filter_func`
+async fn search_package_by_filter<F, QF>(
     package: &PackageName,
-    repo_data: Arc<IndexMap<(Channel, Platform), SparseRepoData>>,
+    all_package_names: Vec<String>,
+    repodata_query_func: QF,
     filter_func: F,
 ) -> miette::Result<Vec<RepoDataRecord>>
 where
     F: Fn(&str, &PackageName) -> bool,
+    QF: FnOnce(Vec<MatchSpec>) -> BoxFuture<'static, Result<Vec<RepoData>, GatewayError>>,
 {
-    let similar_packages = repo_data
+    let similar_packages = all_package_names
         .iter()
-        .flat_map(|(_, repo)| {
-            repo.package_names()
-                .filter(|&name| filter_func(name, package))
+        .filter(|&name| filter_func(name, package))
+        .cloned()
+        .collect_vec();
+
+    // Transform the package names into `MatchSpec`s
+
+    let specs: Vec<MatchSpec> = similar_packages
+        .iter()
+        .map(|name| {
+            let package_name = PackageName::new_unchecked(name);
+            MatchSpec::from(package_name)
         })
-        .unique()
-        .collect::<Vec<&str>>();
+        .collect();
 
-    let mut latest_packages = Vec::new();
+    let repos: Vec<RepoData> = repodata_query_func(specs).await.into_diagnostic()?;
 
-    // search for `similar_packages` in all platform's repodata
-    // add the latest version of the fetched package to latest_packages vector
-    for package in similar_packages {
-        let mut records = Vec::new();
+    let mut latest_packages: Vec<RepoDataRecord> = Vec::new();
 
-        for repo in repo_data.values() {
-            records.extend(
-                repo.load_records(&PackageName::new_unchecked(package))
-                    .into_diagnostic()?,
-            );
-        }
+    for repo in repos {
+        // sort records by version, get the latest one of each package
+        let records_of_repo: HashMap<String, RepoDataRecord> = repo
+            .into_iter()
+            .sorted_by(|a, b| a.package_record.version.cmp(&b.package_record.version))
+            .rev() // Reverse the iterator so that the latest version comes first
+            .fold(HashMap::new(), |mut acc, record| {
+                acc.entry(record.package_record.name.as_normalized().to_string())
+                    .or_insert(record.clone());
+                acc
+            });
```

Comment:
Nice. I think you can also do:

```suggestion
		.map(|r| (record.package_record.name.as_normalized().to_string(), record))
		.collect()
```